protocol/serialization-types/src/type_annotations.rs

```diff
@@ -1,19 +1,49 @@
 // Copyright 2020 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0
 
+use std::io::Read;
+
+use serde::{Deserialize, Serialize};
+
 /// This trait annotates a given type its corresponding bin-prot serialization type,
-pub trait BinProtSerializationType: Sized {
+pub trait BinProtSerializationType<'de>: Sized {
     /// The corresponding serialization type for bin-prot format
     /// Self type can be used here to indicate no special convertion is needed
     /// TODO: Add default value when the feature lands on stable rust
-    type T: From<Self> + Into<Self>;
+    type T: From<Self> + Into<Self> + Serialize + Deserialize<'de>;
+
+    /// Construct from binprot bytes reader
+    fn try_from_binprot(rdr: impl Read) -> Result<Self, bin_prot::error::Error> {
+        let binprot: Self::T = bin_prot::from_reader(rdr)?;
+        Ok(binprot.into())
+    }
+
+    /// Convert into binprot byte vec
+    fn try_into_binprot(self) -> Result<Vec<u8>, bin_prot::error::Error> {
+        let binprot: Self::T = self.into();
+        let mut bytes = Vec::new();
+        bin_prot::to_writer(&mut bytes, &binprot)?;
+        Ok(bytes)
+    }
 }
 
 /// This trait annotates a given type its corresponding json serialization type,
 /// and provide utility functions to easily convert between them
-pub trait JsonSerializationType: Sized {
+pub trait JsonSerializationType<'de>: Sized {
     /// The corresponding serialization type for json format
     /// Self type can be used here to indicate no special convertion is needed
     /// TODO: Add default value when the feature lands on stable rust
-    type T: From<Self> + Into<Self>;
+    type T: From<Self> + Into<Self> + Serialize + Deserialize<'de>;
```

Comment:
Doesn't `From<Self>` imply `Into<Self>` ? do we need `Into<Self>`?

Reply:
I believe From<Self> imply Into<Other>

Reply:
Yeah, just checked, From<T> can also be a lossy conversion. Need to have both